docs/developers/plugin-development.md

```diff
@@ -163,6 +167,106 @@ class MyAwesomePlugin {
+| KEYS | Either an array of keys, or the string `"ALL_KEYS"` or `"NO_KEYS"`, indicating their respective inclusion/exclusion criterea. | `["f", "j"]` |
+| SELECT | A list of strings that a developer can choose between as a parameter. | `["cm", "px", "em"]` |
+| HTML_STRING | A string with HTML markup. | `"<p>This is the prompt.</p>"` |
+| IMAGE | A string that contains the path to an image file. | `"my_image.jpg"` |
+| AUDIO | A string that contains the path to an audio file. | `"my_sound.mp3"` |
+| VIDEO | A string that contains the path to a video file. | `"my_video.mp4"` |
+| OBJECT | A general JSON object (key-value pairs). | `{ rt: 350, response: "hello!", correct: true }` |
+| COMPLEX | A JSON object that one can specify nested parameters for. | `{ rt: 350, response: "hello!", correct: true }` |
+| TIMELINE | A jsPsych timeline object with trials. | `[{ type: jsPsychKeyboardResponse, stimulus: 'my_image.jpg }]` |
+
+Within each parameter, you may also specify if it is an array of the specific type. For example, a parameter that requires a list of button labels would be described as:
+
+```js
+const info = {
+  // ...
+  parameters: {
+    /** The labels to be displayed on each button. */
+    labels: {
+      type: ParameterType.STRING,
+      array: true,
+      default: ["Pause", "Play", "Continue"]
+    }
+  },
+  // ...
+}
+```
+
+Specific parameter types also have their own special markup. For `ParameterType.SELECT`, you specify the options one can choose with an `options` field, and then the `default` field must be within that field.
+
+```js
+const info = {
+  // ...
+  parameters: {
+    /** The units of measure used to display the length and width of the stimulus. */
+    units: {
+      type: ParameterType.SELECT,
+      options: ["em", "px", "vh", "vw"],
+      default: "px"
+    }
+  },
+  // ...
+}
+```
+
+For `ParameterType.COMPLEX`, we may specify the underlying fields in the object with the `nested` field. This acts in the same way as us defining parameters regularly, only we are now just delineating the fields within the object itself.
+
+```js
+const info = {
+  // ...
+  parameters: {
+    /** Where to display the location of the stimuli. */
+    locations: {
+      type: ParameterType.COMPLEX,
+      array: true,
+      default: undefined,
+      nested: {
+        /** The x-coordinate of the stimulus, in the units from the `units` field. */
+        x: {
+          type: ParameterType.INT
+        },
+        /** The y-coordinate of the stimulus. */
+        y: {
+          type: ParameterType.INT
+        }
+      }
+    }
+  },
+  // ...
+}
+```
+
+For more complicated scenarios, typically when handling data generated from an arbitrary function or user input, where we have a general idea of what data type it could produce, we may also specify multiple types of data. As an example, if we know we'll get either some number (integer or float) or a string from a field, we can specify it as such:
```

Comment:
this shows up primarily in the survey suite of plugins, where the data is specified as multiple types, but nowhere else. wondering if it's an issue with the implementation of survey as all `data.response` are sent as objects regardless of the type in the object, so something to look a bit deeper at
```js
response: { // survey-html-form implementation
          type:
            ParameterType.STRING |
            ParameterType.INT |
            ParameterType.FLOAT |
            ParameterType.BOOL |
            ParameterType.OBJECT,
},
```

Reply:
I think you are right that in the survey plugins the data comes back as objects. So I'm still a bit puzzled why there would be different properties. Maybe this was just an oversight when the data types were being added over the summer?

Reply:
i believe it is, as there's also no support for this in parameters. but looking a bit deeper, it seems that these bits only really persist in the data object, which isn't vetted in the same way as the parameters (or rather, not vetted at all). i'm removing the section on documentation but i believe that changing the survey data to just be of `ParameterType.OBJECT` has no effect on the end code cause nothing is measuring the last parts of the data, but i could be wrong. @vzhang03 thoughts, since you worked on this?

Reply:
i think in this case we can just remove use object to clear up any confusion, especially for people looking to write plugins based off of current code. @jodeleeuw i think we're all good to merge this, and i'll write a PR to remove the erroneous data tagging in our survey plugins